### DISCUSSION TITLE
Faster way to iterate with custom scripts?

### DISCUSSION THREAD

**Discussion by justyns** [netbox-community/netbox#14593] [Help Wanted!]
Hello!

I recently started converting some scripts that use pynetbox to custom scripts that can run inside of netbox.   Apologies if I missed it, but is there a faster method to test changing the script versus this?

1. Change code
2. Commit and push code
3. Go to data source job and click sync
4. Go to custom script and run

I do have a repo added as a datasource using git.   However sometimes I want to test something simple like just checking if a property is available on a particular object and it feels like several steps to get from "code" -> "run code".  

Is there a way to run the custom script locally, or some other way to make this cycle a little faster? 

And bonus question, is there a way to schedule automatic updates of data sources?

**Comment by candlerb** [comment]
A good way to "test something simple like just checking if a property is available on a particular object" is to use [nbshell](https://docs.netbox.dev/en/stable/administration/netbox-shell/):

```
# /opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py nbshell
...
help(Device)
d = Device.objects.get(name="foo")
dir(d)
d.id
d.name
```

I'm fairly sure that if you upload the script via the web interface, it's OK to edit it directly under `/opt/netbox/netbox/scripts/`. I don't use the git data sources though, so I don't know if there's a git checkout you can access (and if there is, once you've finished testing you'll probably want to reset the checkout to avoid conflicts when you next pull)

**Reply by justyns** [reply]
Thanks!  I didn't know about nbshell, playing around with it a little looks like exactly what I need though.  

> I'm fairly sure that if you upload the script via the web interface, it's OK to edit it directly under /opt/netbox/netbox/scripts/

I'll give this a try too.   If it is safe, I'd be fine with having a test version of the script that I'm editing manually so that it doesn't conflict with the version from the git data source.

**Reply by llamafilm** [reply]
On my system `/opt/netbox/netbox/scripts` is empty, how is that meant to be used?  I put my scripts in a local data source, so after editing the file I do:
1. Go to data source job and click sync
2. Go to custom script and run

Not bad, but I wish there were some way to run the code directly in the shell, without having to sync the data source.

**Reply by candlerb** [reply]
You can upload a script via the web interface (Customation > Scripts > +, File Upload). This will put it into /opt/netbox/netbox/scripts/ and also create the necessary database entry. Once done, you can edit it *in situ* in the filesystem.

**Comment by zamanrahimi** [comment]
Hello Everyone, 
Question: Both solutions are attempting to load UN/LOCODE or location codes in a NetBox custom script.
However, since UN/LOCODE's records are around 20k, it is extremely slow.
The reason is that all the records are loaded into memory. So, does anyone have a solution for the performance issue?
Has anyone experienced the same issue?


    # solution 1 
    # in this solution, unlocode1 which is a custom field choic, directly loaded 
    field_name = ChoiceVar(
        description="Field description",
        choices=CustomFieldChoiceSet.objects.get(name="unlocode1").choices,
         choices=choices
        required=True,
        )


    # solution 2 
    # in this solution, unlocode1 which is a custom field choice, first; loaded in unlocode2 which is a custom field
    # then custom field untest2 is loaded bellow    
    custom_field_name = "unlocode2" # unlocode2 is a custom field which loads a unlocode1 which is custom chioice field 
    custom_field = CustomField.objects.get(name=custom_field_name)
    choices = custom_field.choices

    field_name = ChoiceVar(
        description="Field description",
        choices=choices,
        required=True,
        )
